### PR TITLE
Markdown parser updates

### PR DESCRIPTION
**Pull request by BrutuZ**
Parsed after the markdown format since it has the added benefit of avoiding its closing brackets at the end of the URL

### CODE REVIEW

File: proxy/source/markdown_parser.py [hunk 1]
```diff
@@ -58,12 +58,18 @@ def _convert_crlf(input_str: str) -> str:
...
 def _parse_links(input_str: str) -> str:
-    return re.sub(
+    input_str = re.sub(
         r"\[([\w\W]+?)\]\(([\w\W]+?)\)",
         r'<a href="\2">\1</a>',
         input_str,
         flags=re.MULTILINE,
     )
+    return re.sub(
+        r"(?<!href=\")(https?:\/\/[-a-zA-Z0-9._~:/?#@!$&()*+,;=%]+')",
```
Comment (funkyhippo) [inline proxy/source/markdown_parser.py]:
```suggestion
        r"(?<!href=\")(https?:\/\/[-a-zA-Z0-9._~:/?#@!$&()*+,;=%]+)",
```

I think that dangling `'` may have been breaking the regex.

**Review by funkyhippo — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Let me know if I missed something regarding the regex, but thanks for tightening it up rather than hitting it with the ol' `\w\W`.

**Comment by BrutuZ** [comment]
It shouldn't break in pure Python, I did test it in the console to make sure it was matching as expected and had no issues.
Maybe escape the single quote too rather than remove it? It is a valid URL character after all, and not that unusual since it's often used to enclose values in parameters

**Comment by funkyhippo** [comment]
Break may have been the wrong word. What's your expected input?

```python
>>> import re
>>> def _parse_links(input_str: str) -> str:
...     input_str = re.sub(
...         r"\[([\w\W]+?)\]\(([\w\W]+?)\)",
...         r'<a href="\2">\1</a>',
...         input_str,
...         flags=re.MULTILINE,
...     )
...     return re.sub(
...         r"(?<!href=\")(https?:\/\/[-a-zA-Z0-9._~:/?#@!$&()*+,;=%]+')",
...         r'<a href="\1">\1</a>',
...         input_str,
...         flags=re.MULTILINE,
...     )
...
>>> test_str = """
... https://example.com
... [example](https://example.com)
... https://example.com'
... """
>>> print(_parse_links(test_str))

https://example.com
<a href="https://example.com">example</a>
<a href="https://example.com'">https://example.com'</a>

```

Did you intend to only support links that were suffixed with `'`? It's outside the character set.

**Comment by BrutuZ** [comment]
> Did you intend to only support links that were suffixed with '? It's outside the character set.

That... was an oversight between converting the expression enclosure from single to double-quotes in the browser editor 😅

**Comment by BrutuZ** [comment]
Added some other minor changes, doesn't seem like I broke anything 😅
[image](https://cdn.discordapp.com/attachments/986447226844885062/1102787494006759424/image.png)

**Comment by BrutuZ** [comment]
Anything else holding this and #20 ?
I have the changes from both PRs ([and a few more](https://github.com/subject-f/cubarimoe/compare/develop...BrutuZ:cubarimoe:webtest)) running on a forked instance with no issues so far

**Comment by funkyhippo** [comment]
No blockers at a glance -- sorry for the slow review, I've just been incredibly busy with irl.

I'll block out some time this weekend to take a closer look and/or merge.

File: proxy/source/markdown_parser.py [hunk 2]
```diff
@@ -83,14 +93,14 @@ def _parse_strong_emphasis(input_str: str) -> str:
 
 def _parse_em_emphasis(input_str: str) -> str:
     return "\n".join(
-        re.sub(r"(?:\*|\_)([\w]+?[\w\W]+?[\w]+?)(?:\*|\_)", r"<em>\1</em>", l)
+        re.sub(r"[*_](.+?)[*_]", r"<em>\1</em>", l)
```
Comment (funkyhippo) [inline proxy/source/markdown_parser.py]: I'm generally fine with these simplifications, but it does stray a bit further from Gruber's original markdown spec:

> But if you surround an * or _ with spaces, it’ll be treated as a literal asterisk or underscore.

([source](https://daringfireball.net/projects/markdown/syntax#:~:text=But%20if%20you%20surround%20an%20*%20or%20_%20with%20spaces%2C%20it%E2%80%99ll%20be%20treated%20as%20a%20literal%20asterisk%20or%20underscore.))

File: proxy/source/markdown_parser.py [hunk 3]
```diff
@@ -58,22 +54,36 @@ def _convert_crlf(input_str: str) -> str:
...
+            input_str,
+        )
+    return input_str
 
 
 def _parse_strong_emphasis(input_str: str) -> str:
     return "\n".join(
         re.sub(
-            r"(?:\*\*|\_\_)([\w]+?[\w\W]+?[\w]+?)(?:\*\*|\_\_)",
+            r"[*_]{2}(.+?)[*_]{2}",
```
Comment (funkyhippo) [inline proxy/source/markdown_parser.py]: I'm a bit iffy on this simplification, since it allows this to be parsed as bolded text:

```
*_asdf**
```

File: proxy/source/markdown_parser.py [hunk 4]
```diff
@@ -58,22 +54,36 @@ def _convert_crlf(input_str: str) -> str:
...
+        r'<a href="\1" target="_blank" rel="nofollow noreferrer noopener">\1</a>',
         input_str,
-        flags=re.MULTILINE,
+        flags=re.MULTILINE|re.IGNORECASE,
     )
 
 
 def _parse_headers(input_str: str) -> str:
-    return re.sub(r"^#+ +([\w\W]+?)\n", r"<h3>\1</h3>\n", input_str, flags=re.MULTILINE)
+    search = re.finditer(r"^(#{1,5}) (.+)$", input_str, re.MULTILINE)
```
Comment (funkyhippo) [inline proxy/source/markdown_parser.py]: I think there may be a behaviour change here. The original regex captured and absorbed all the whitespace between the hash and the next valid character:

#### Original

```python
>>> text = """
... #   some header with multiple spaces
... """
>>> _parse_headers(text)
'\n<h3>some header with multiple spaces</h3>\n'
```

#### New

```python
>>> text = """
... #   some header with multiple spaces
... """
>>> _parse_headers(text)
'\n<span style="font-size: 1.8em;">  some header with multiple spaces</span>\n'
```

The spec is ambiguous about this though, so not a blocker.

File: proxy/source/markdown_parser.py [hunk 5]
```diff
@@ -9,7 +9,7 @@
 
 # The supported tags are:
 # - URLs
-# - Headers (atx-style, parses to a single level)
+# - Headers (span blocks with font-size, parses up to 4 levels)
```
Comment (funkyhippo) [inline proxy/source/markdown_parser.py]: nit: we should keep `atx-style` in the comment since we don't support Setex style.

**Review by funkyhippo — COMMENTED** [review]
Sorry for the slow review -- I left some comments.

**Comment by BrutuZ** [comment]
Should have addressed all notes 🤞

File: proxy/source/markdown_parser.py [hunk 6]
```diff
@@ -58,22 +54,36 @@ def _convert_crlf(input_str: str) -> str:
...
+            input_str,
+        )
+    return input_str
 
 
 def _parse_strong_emphasis(input_str: str) -> str:
     return "\n".join(
         re.sub(
-            r"(?:\*\*|\_\_)([\w]+?[\w\W]+?[\w]+?)(?:\*\*|\_\_)",
+            r"(?<!\\)\*\*(\w.*?)(?<!\\)\*\*|(?<!\\)__(\w.*?)(?<!\\)__",
```
Comment (funkyhippo) [inline proxy/source/markdown_parser.py:86]: Nice, I'm a fan of also handling the escaped character.

I think you're handling the first character being a space, but not the last?

```python
>>> text = """
... **asdf**
... ** asdf**
... **asdf **
... ** asdf **
... """
>>> print(_parse_strong_emphasis(text))

<strong>asdf</strong>
** asdf**
<strong>asdf </strong>
** asdf **
```

It's fine though, I'm not going to block on this (especially since the original impl was flawed as it only allowed `\w` which ignores symbols as valid characters).

**Review by funkyhippo — APPROVED** [review]
Sorry for the long back-and-forth, I appreciate the iteration. A set of tests would be useful so that we can capture all the edge cases and expected behaviour.

Going to merge this as-is, thanks so much for the contribution!

**Comment by BrutuZ** [comment]
I couldn't figure an elegant way to filter trailing spaces as well. While another \w at the end might handle it, it would also make it impossible to format single characters like **A** letter or _1_ digit.
Turns out a space with the {0} quantifier doesn't work as a negative 🙁
